### ISSUE TITLE
Code of Loss (objective) function

### ISSUE THREAD

**Issue by irleader** [keio-bioinformatics/mxfold2#1]
Hi Dr Kengo Sato,

I have been reading your paper as well as trying to understand this github repo code. There has been several doubts I would like to ask you about. For this issue, I will ask about the corresponding code of loss function.

In the paper, the objective function is:

![Screen Shot 2021-02-09 at 7 48 55 pm](https://user-images.githubusercontent.com/37386488/107338622-d9f00780-6b0f-11eb-86ca-801a701b4671.png)

I find the corresponding python loss function code from loss.py:
```
class StructuredLossWithTurner(nn.Module):
    def __init__(self, model, loss_pos_paired=0, loss_neg_paired=0, loss_pos_unpaired=0, loss_neg_unpaired=0, 
                l1_weight=0., l2_weight=0., sl_weight=1., verbose=False):
        super(StructuredLossWithTurner, self).__init__()
        self.model = model
        self.loss_pos_paired = loss_pos_paired
        self.loss_neg_paired = loss_neg_paired
        self.loss_pos_unpaired = loss_pos_unpaired
        self.loss_neg_unpaired = loss_neg_unpaired
        self.l1_weight = l1_weight
        self.l2_weight = l2_weight
        self.sl_weight = sl_weight
        self.verbose = verbose
        from .fold.rnafold import RNAFold
        from . import param_turner2004
        if getattr(self.model, "turner", None):
            self.turner = self.model.turner
        else:
            self.turner = RNAFold(param_turner2004).to(next(self.model.parameters()).device)

    def forward(self, seq, pairs, fname=None):
        pred, pred_s, _, param = self.model(seq, return_param=True, reference=pairs,
                                loss_pos_paired=self.loss_pos_paired, loss_neg_paired=self.loss_neg_paired, 
                                loss_pos_unpaired=self.loss_pos_unpaired, loss_neg_unpaired=self.loss_neg_unpaired)
        ref, ref_s, _ = self.model(seq, param=param, constraint=pairs, max_internal_length=None)
        with torch.no_grad():
            ref2, ref2_s, _ = self.turner(seq, constraint=pairs, max_internal_length=None)
        l = torch.tensor([len(s) for s in seq], device=pred.device)
        loss = (pred - ref) / l
        loss += self.sl_weight * (ref-ref2) * (ref-ref2) / l
        if self.verbose:
            print("Loss = {} = ({} - {})".format(loss.item(), pred.item(), ref.item()))
            print(seq)
            print(pred_s)
            print(ref_s)
        if loss.item()> 1e10 or torch.isnan(loss):
            print()
            print(fname)
            print(loss.item(), pred.item(), ref.item())
            print(seq)

        if self.l1_weight > 0.0:
            for p in self.model.parameters():
                loss += self.l1_weight * torch.sum(torch.abs(p))

        # if self.l2_weight > 0.0:
        #     l2_reg = 0.0
        #     for p in self.model.parameters():
        #         l2_reg += torch.sum((self.l2_weight * p) ** 2)
        #     loss += torch.sqrt(l2_reg)

        return loss
```

From the python code only, I am not able to match to the function from paper. I can see both L1 and L2 regularization terms, but I am not able to find the code for structured hinge loss function (margin term and max function), I suppose f(x, y) from paper means the score (first return item) from self.model. I assume that the max margin part is embedded in self.model which is in the C++ code part. From the paper, you are using MixedFold, can you please kindly point out which part and lines of C++ code are the max margin function?

**Comment by satoken** [comment]
Hi,
Thank you for your interest in mxfold2.

I employed a technique called the loss-augmented inference [Tsochantaridis et al. 2005] to calculate the structured hinge loss, as I wrote in our paper.
Two hyperparameters, loss_pos_paired and loss_neg_paired used by \delta(y, \hat{y}), are passed to C++ code, and used at mxfold2/src/fold/zuker.cpp:70.
These penalize the folding score to maximize the margin between the correct structure and others.

Thanks,

**Comment by irleader** [comment]
Hi,

Thanks for your information.

After studying the paper you provided and online courses of Structured Learning, I understand why you are using SSVM and why the loss function is defined like this. 

1. Also, according to you and the paper, the two hyper parameters  δFP = 0.005 and δFN = 0.5 should be loss_pos_paired and loss_neg_paired from loss.py. They are passed to C++code and used at zuker.cpp:70. But I am not very sure what the two functions at zuker.cpp:69 and zuker.cpp:70 do (opts.make_constraint() and opts.make_penalty() )? Can you explain this a bit?

2. You talked about "loss-augmented inference", but I am not able to find information about it either from referenced paper or online. In Structured SVM, the inference problem can usually be solved by dynamic programming like Viterbi. I see from your C++code that you also used Viterbi. Do you use Viterbi to solve max [f(x, yˆ) + ∆(y, yˆ)] ?

Thanks a lot in advance.

**Comment by satoken** [comment]
Both of these questions are about the structured hinge loss. This PDF may answer your questions:
https://github.com/keio-bioinformatics/mxfold/raw/master/suppl.pdf

Thanks,
